examples/usercall-extension-bind/runner/src/main.rs

```diff
@@ -0,0 +1,331 @@
+            self.fuse("Protocol header signature mismatch")?
+        }
+
+        let (min_len, peer_addr_len) = match (vercmd, family) {
+            (V2CMD_LOCAL, FAMILY_UNSPEC) => (0, 0),
+            (V2CMD_PROXY, FAMILY_TCP4) => (12, 4),
+            (V2CMD_PROXY, FAMILY_TCP6) => (36, 16),
+            _ => self.fuse("Unsupported version/command/family")?
+        };
+
+        if addr_len < min_len {
+            self.fuse("Unexpected address length received")?
+        }
+
+        self.read_upto(SIZE_HEADER + addr_len as usize)?;
+
+        let port = u16::from_be_bytes(self.buf[self.buf.len()-2..].try_into().unwrap());
+        let addr = self.buf[SIZE_HEADER..SIZE_HEADER + peer_addr_len].to_owned();
+        match family {
+           FAMILY_TCP4 => {
+                    let mut arr = [0 as u8; 4];
+                    for (idx, element) in addr.into_iter().enumerate().take(arr.len()) {
+                        arr[idx] = element;
+                    }
+                    Ok(Some(SocketAddr::new(arr.into(), port)))
+                 },
+            FAMILY_TCP6 => {
+                    let mut arr = [0 as u8; 16];
+                    for (idx, element) in addr.into_iter().enumerate().take(arr.len()) {
+                        arr[idx] = element;
+                    }
+                    Ok(Some(SocketAddr::new(arr.into(), port)))
+                 },
+            FAMILY_UNSPEC => {
+                 Ok(None)
+            },
+            _ => {
+                   self.fuse("Unsupported addr size")
+                  }
+        }
+    }
+
+    fn fuse<T>(&mut self, message: &'static str) -> IoResult<T> {
+        eprintln!("Bailing out of proxy protocol handler: {}", message);
+        self.fused = true;
+        Err(Error::new(ErrorKind::InvalidData, message))
+    }
+
+    pub fn advance(&mut self) -> IoResult<Option<SocketAddr>> {
+        if self.fused {
+            self.fuse("(repeat call)")?
+        }
+
+        self.advance_stream()
+    }
+    pub fn new(stream : TcpStream) -> IoResult<(Self, Option<SocketAddr>)> {
+        let mut ret = ProxyState {
+            stream        : stream,
+            fused         : false,
+            buf           : vec![],
+            position      : 0,
+        };
+        let peer_addr = ret.advance()?;
+        Ok((ret, peer_addr))
+    }
+}
+
+struct ProxyProtocol {
+    listener      : TcpListener,
+    pub local_address : String,
+}
+
+impl ProxyProtocol {
+    fn new(addr: &str)-> IoResult<Self> {
+        TcpListener::bind(addr).map(|listener| {
+                let local_address = match listener.local_addr() {
+                    Ok(local_address) => local_address.to_string(),
+                    Err(_) => "error".to_string(),
+                };
+                ProxyProtocol {listener, local_address}
+        })
+    }
+}
+
+impl SyncListener for ProxyProtocol {
+    fn accept(&self) -> IoResult<(Box<SyncStream>, Box<ToString>, Box<ToString>)> {
```

Comment:
Since we're basically guaranteed to have stripped off the proxy protocol by the time this function returns, I think the logic in this module can be significantly simplified: we don't need to do the fusing or separate state tracking in `ProxyState`. I think all that's needed is a `fn read_proxy_protocol_header(&mut TcpStream) -> Result<Option<(SocketAddr, SocketAddr)>>`. After the proxy protocol has been read, you can simply return the `Box<TcpStream>` from this function. Make sure to also get the local address from the proxy protocol.

Reply:
Thanks @jethrogb,  I have simplified the code to remove `ProxyState`.
Reading stream's local address in app would give an error for local profile, where we were filling `local_addr` as `Some("Err".to_string())`.
As of now, I am filling in `"0.0.0.0:0"` would that be ok?, In that case should I make `Option<(SocketAddr, SocketAddr)>` as `(SocketAddr, SocketAddr)`.

Reply:
For local, I think it's best to use the TCP stream's actual addresses.

Reply:
Thanks @jethrogb 
Recent push returns TCP stream's actual addresses for local profile.